testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala

```diff
@@ -1,11 +1,16 @@
 package edu.illinois.cs.testrunner.data.framework
 
+import edu.illinois.cs.testrunner.testobjects.GeneralTestClass
+import edu.illinois.cs.testrunner.testobjects.JUnitTestCaseClass
+import edu.illinois.cs.testrunner.testobjects.JUnitTestClass
+import edu.illinois.cs.testrunner.testobjects.JUnit5TestClass
 import org.apache.maven.project.MavenProject
-
+import java.lang.annotation.Annotation
+import java.lang.reflect.Modifier
+import java.util.List;
 import scala.collection.JavaConverters._
-
-sealed abstract class TestFramework
-case object JUnit extends TestFramework
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
 
 object TestFramework {
     def testFramework(project: MavenProject): Option[TestFramework] = {
```

Comment:
It seems like this method now returns the relevant TestFramework as long as there is just one, but right now it is only relevant for JUnit 4. Should it be modified to also check if junit-jupiter is there and return JUnit5?

Reply:
I agree that this method doesn't look right. This method has only one output `TestFramework`, but there are two possible frameworks.

If the maven project has two frameworks (J4 and J5), what is the expected output of this method? It seems that my newly added method `getListOfFrameworks` can cover this method, which means that we don't need this method anymore.

Reply:
So from what I understand, the TestFramework is used to determine what runner or test executor to use to run tests. The TestFramework is also used to determine what are the tests to run, which is based off of annotations. Is it possible for JUnit 5 to run JUnit 4 style tests? If a project actually does have both JUnit 4 and JUnit 5 tests, does it utilize JUnit 5 to actually run the JUnit 4 tests (i.e., the dependency on JUnit 4 is just for compilation purposes)?

Basically, if JUnit 5 can run JUnit 4 tests, then perhaps the code for JUnit 5 could be updated to also service JUnit 4 tests in case both are present?

Reply:
It's possible for JUnit 5 framework to run JUnit 4 style tests, and I have an [implementation](https://github.com/luoos/testrunner/commit/6d21eb7f0979b2c2f50d278a7b1a3672d1e57c8e) for it. But there is an issue. Using J5 framework to run J4 tests needs a special runtime dependency - [junit-vintage-engine](https://junit.org/junit5/docs/current/api/org.junit.vintage.engine/module-summary.html). This issue is pretty tricky. And I discussed it with the professor.

In short, when the testrunner starts a new process to run tests, we need to add the jar file path of the `junit-vintage-engine` dependency to `classpath` (-cp) parameters. However, it's a bit tricky to do that. One simple way is to modify the target project's pom.xml, adding the `junit-vintage-engine` as a dependency. But this method is verbose and error-prone.

Combining that we don't have full control of test order with J5 API (if not use the "separate" method as a workaround), it would be safer, for now, to keep using the original J4 executor to run J4 tests, and the J5 executor to run J5 tests.

Reply:
I assume you mean when there are a mix of JUnit 4 and 5 tests then you would need the junit-vintage-engine dependency to run them (because if there was only JUnit 4 tests and no dependency on junit-jupiter, then you wouldn't need to have JUnit 5 run the JUnit 4 tests). However, if there is a mix, does the target project need to have junit-vintage-engine as a dependency already to even run those tests? Should that dependency already be there in such a case?

Concerning using JUnit 5 API to run JUnit 4 tests, is it that running them in alphabetical order a built-in requirement? In other words, if one would to normally run JUnit 4 tests through JUnit 5 in a project, does the specification for JUnit 5 say that those tests would be run in alphabetical order?

Reply:
For a mixed project, it typically has 1) `junit` dependency to run J4 tests (*J4 framework to J4 tests*), 2) `junit-jupiter` dependency to run J5 tests (*J5 framework to run J5 tests*). It doesn't need `junit-vintage-engine`. But what we're trying to do here, is to use J5 API to run J4 tests, that's quite different. So we need to add the `junit-vintage-engine` to the target project.

"running them in alphabetical order" is what I observed with my [demo project](https://github.com/luoos/junit5-testrunner-demo), I think that's built-in behavior.

Reply:
How about we change this method to check for which dependency exists and returns the one that does exist, and if both exist then either throw some exception about there being multiple relevant dependencies or returning Option.empty. We should leave a note about this issue that this method assumes only no mix of dependencies and one should use getListOfFrameworks for the more general case. We can look into this more later if we end up needing more support for this mix.

From what I understand, while one can run both JUnit 4 and JUnit 5 tests by having both dependencies, the "correct", recommended way is to include that junit-vintage-engine so JUnit 5 can run JUnit 4 tests. I suppose we need to support even this other way with both dependencies for the projects that don't play so nice, but I don't think it's necessarily wrong to always assume junit-vintage-engine is not there.

Reply:
Updated the testFramework method as required.

I'm not sure about the "correct" practice for a J4 and J5 mixed project. But for J4-only project, it definitely doesn't contain `junit-vintage-engine` dependency. So if we want to use J5 API to run J4 tests, we still need to add that dependency. So we can't simply assume that dependency exists.

Reply:
Thanks for making the change. This should be good for now to merge.

To elaborate a bit more, yes, if the project only has JUnit 4 (where it doesn't even have any dependency on junit-jupiter), then we should for sure run using the JUnit 4 API. I am not suggesting that we only use JUnit 5 to run tests, regardless of present dependencies. If the project has both JUnit 4 and JUnit 5 tests, then I think it is recommended that the project use junit-vintage-engine to run the JUnit 4 style tests.